### ISSUE TITLE
[Potential Bug] SAE Steering Not Applied to Each Decoded Token in VLLM

### ISSUE THREAD

**Issue by Peter-Fy** [AIRI-Institute/SAE-Reasoning#3]
It appears that the feature steering operation is not being applied to every decoded token during generation. **I would like to know how to modify the code to ensure the steering operation is properly applied to each token during generation.**

Following are how I found the decoding is not working as expected:

## Investigation 1

I modified the `get_intervention_hook` in [sae_utils.py](https://github.com/AIRI-Institute/SAE-Reasoning/blob/main/evaluation/lm-evaluation-harness/lm_eval/sae_utils.py) to track the total number of steering operations:

```python
steering_times = 0
def get_intervention_hook(
    sae: SAE,
    feature_idx: int,
    max_activation: float = 1.0,
    strength: float = 1.0,
):
    def hook_fn(module, input, output):
        global steering_times
        # ... same as before
        print("\n" + "="*100)
        steering_times += 1
        print(f"steering_times: {steering_times}")
        print("="*100 + "\n")

        if torch.is_tensor(output):
            return activations_hat
        else:
            return (activations_hat,) + output[1:] if len(output) > 1 else (activations_hat,)

    return hook_fn
```
The results show that **the steering operation is only applied 32,769 times for all 30 questions in AIME24**, which seems insufficient for token-by-token intervention.

## Investigation 2

Further testing with the following code reveals that **the steering only affects the first decoded token for each generation function call**:

``` python
from vllm import LLM, SamplingParams
from transformers import AutoTokenizer
import os
from sae_utils import add_hooks, get_intervention_hook, get_clamp_hook
from sae_lens import SAE
from sae_utils import add_hooks, get_intervention_hook, get_clamp_hook

# load llm
os.environ["VLLM_USE_V1"] = '0'
model = 'deepseek-ai/DeepSeek-R1-Distill-Llama-8B'
tokenizer = None
llm = LLM(model=model, max_model_len=32768, gpu_memory_utilization=0.8,trust_remote_code=False) 
sampling_params = SamplingParams(temperature=0.6,top_p=0.95,max_tokens=32768)

# load sae
sae, cfg_dict, sparsity = SAE.from_pretrained(
    release="andreuka18/deepseek-r1-distill-llama-8b-lmsys-openthoughts",  # see other options in sae_lens/pretrained_saes.yaml
    sae_id="blocks.19.hook_resid_post",  # won't always be a hook point
    device="cuda",
)

# test
messages = [
    {"role": "user", "content": "How many r’s are in the word strawberry?"}
]
tokenizer = llm.get_tokenizer()
text = tokenizer.apply_chat_template(
    messages,
    tokenize=False,
    add_generation_prompt=True
)

lm_model = llm.llm_engine.model_executor.driver_worker.model_runner.model
sampling_params = SamplingParams(temperature=0,top_p=1,max_tokens=1024)
sae_hooks = []
feature_idx = 46379
strength = 2
max_activation = 5.840
direction = sae.W_dec[feature_idx].clone()
sae_hooks.append(
    (
        lm_model.model.layers[sae.cfg.hook_layer],
        get_clamp_hook(direction, max_activation, strength)
    )
)

with add_hooks([], sae_hooks):
    outputs = llm.generate([text], sampling_params)

for output in outputs:
    prompt = output.prompt
    generated_text = output.outputs[0].text
    print(generated_text)
```

**Comment by avgalichin** [comment]
Hello!

Thank you for the interest in our work.

- Investigation 1: Please note that the `vllm.LLM.generate` has its own efficient batching method, and each forward step is performed in batches. Therefore, when the hook is called once, it is actually called for the batch of activations, each corresponding to a different token. So the correct update for the `steering_times` should be e.g. `steering_times += activations.size(0)`. This should give you the total number of token activations the hook intervened upon.
- Investigation 2: Based on the code you've sent, I see that you do not pass the `enforce_eager = True` in the `vllm.LLM` constructor. This means that `vllm` will build a graph during the initial forward passes, and then reuse it during further calls. The hook then will indeed be called only once during the forward step corresponding to the graph construction. To enable hook calling at each step, you should set `enforce_eager = True`. This has been also discussed in [(#4084)](https://github.com/vllm-project/vllm/issues/4084).

In the `lm_eval.models.vllm_causallms.py` we set `enforce_eager = True` if SAE is being initialized - please see [here](https://github.com/AIRI-Institute/SAE-Reasoning/blob/main/evaluation/lm-evaluation-harness/lm_eval/models/vllm_causallms.py#L132).

I hope my answers solve the questions. If you have additional concerns/questions, please feel free to ask.

**Comment by Peter-Fy** [comment]
Thank you for your prompt and helpful response, which completely solved my problem!

I have one more question regarding the ``evaluation.sh``. I noticed there's a parameter ``sae_max_activation=5.840`` in the configuration. I wander how to obtain this value for other feature indices mentioned in the paper (such as 15136, 17456, and 62777).

**Comment by avgalichin** [comment]
I'm glad the previous response was helpful! 

Regarding your question about the `sae_max_activation` parameter: this is the maximum feature activation value estimated over some small fraction of OpenThoughts-114k dataset.

To find the value by yourself, you can check the official `SAELens` [guide](https://github.com/jbloomAus/SAELens/blob/main/tutorials/tutorial_2_0.ipynb), `Basics: Intervening on SAE Features` section, `find_max_activation` function.

But I understand that this could be hard to implement by yourself; I will try to add the corresponding notebook in the next repository update. For now, here is the `sae_max_activation` values for the features mentioned in the paper:

- **15136**: 3.362
- **17456**: 4.212
- **46379**: 5.840
- **62777**: 4.527

**Comment by Peter-Fy** [comment]
Thank you so much! Your work continues to inspire me, and I deeply appreciate your time and guidance! 😊

**Comment by avgalichin** [comment]
I'm glad I could help, and thank you for the good words!

**Comment by Peter-Fy** [comment]
When I run the evaluation, the generated results remain unchanged each time, even though I have explicitly set the temperature and top_p parameters. I would greatly appreciate any insights you might have regarding what could be causing this unchanging behavior.

Here's my evaluation script:

``` bash
lm_eval --model vllm \
  --model_args pretrained=deepseek-ai/DeepSeek-R1-Distill-Llama-8B,\
    dtype=bfloat16,\
    tensor_parallel_size=1,\
    gpu_memory_utilization=0.85,\
    sae_release=andreuka18/deepseek-r1-distill-llama-8b-lmsys-openthoughts,\
    sae_id=blocks.19.hook_resid_post,\
    sae_feature_idx=46379,\
    sae_strength=2.0,\
    sae_max_activation=5.840 \
  --tasks aime2024_nofigures \
  --batch_size auto \
  --apply_chat_template \
  --output_path deepseek-llama-8b-aime-sae-46379-test \
  --log_samples \
  --gen_kwargs "max_gen_toks=32768,cutoff_token=128014,do_sample=True,temperature=0.6,top_p=0.95"

**Comment by CurryxIaoHu** [comment]
Do authors know how to modify LLM's hidden states during vLLM inference? I have been troubled by this for a long time. I would be grateful if you can give me some help.

**Comment by avgalichin** [comment]
> When I run the evaluation, the generated results remain unchanged each time, even though I have explicitly set the temperature and top_p parameters. I would greatly appreciate any insights you might have regarding what could be causing this unchanging behavior.
> 
> Here's my evaluation script:
> 
> lm_eval --model vllm \
>   --model_args pretrained=deepseek-ai/DeepSeek-R1-Distill-Llama-8B,\
>     dtype=bfloat16,\
>     tensor_parallel_size=1,\
>     gpu_memory_utilization=0.85,\
>     sae_release=andreuka18/deepseek-r1-distill-llama-8b-lmsys-openthoughts,\
>     sae_id=blocks.19.hook_resid_post,\
>     sae_feature_idx=46379,\
>     sae_strength=2.0,\
>     sae_max_activation=5.840 \
>   --tasks aime2024_nofigures \
>   --batch_size auto \
>   --apply_chat_template \
>   --output_path deepseek-llama-8b-aime-sae-46379-test \
>   --log_samples \
>   --gen_kwargs "max_gen_toks=32768,cutoff_token=128014,do_sample=True,temperature=0.6,top_p=0.95"

Can you clarify what do you mean by "unchanged"? Unchanged compared to what?

**Comment by avgalichin** [comment]
> Do authors know how to modify LLM's hidden states during vLLM inference? I have been troubled by this for a long time. I would be grateful if you can give me some help.

You can do this via `enforce_eager = True` and the pytorch hook functionality. The hook should be attached to the initial model which is stored in `vllm.LLM` here: `.llm_engine.model_executor.driver_worker.model_runner.model`

**Comment by Peter-Fy** [comment]
> > When I run the evaluation, the generated results remain unchanged each time, even though I have explicitly set the temperature and top_p parameters. I would greatly appreciate any insights you might have regarding what could be causing this unchanging behavior.
> > Here's my evaluation script:
> > lm_eval --model vllm 
> > --model_args pretrained=deepseek-ai/DeepSeek-R1-Distill-Llama-8B,
> > dtype=bfloat16,
> > tensor_parallel_size=1,
> > gpu_memory_utilization=0.85,
> > sae_release=andreuka18/deepseek-r1-distill-llama-8b-lmsys-openthoughts,
> > sae_id=blocks.19.hook_resid_post,
> > sae_feature_idx=46379,
> > sae_strength=2.0,
> > sae_max_activation=5.840 
> > --tasks aime2024_nofigures 
> > --batch_size auto 
> > --apply_chat_template 
> > --output_path deepseek-llama-8b-aime-sae-46379-test 
> > --log_samples 
> > --gen_kwargs "max_gen_toks=32768,cutoff_token=128014,do_sample=True,temperature=0.6,top_p=0.95"
> 
> Can you clarify what do you mean by "unchanged"? Unchanged compared to what?

When I run this script multiple times with the same parameters, for each AIME problem, I get exactly identical responses across different runs. Since I've set temperature=0.6 and top_p=0.95 with do_sample=True, I would expect some randomness in the generated responses.

**Comment by avgalichin** [comment]
> > > When I run the evaluation, the generated results remain unchanged each time, even though I have explicitly set the temperature and top_p parameters. I would greatly appreciate any insights you might have regarding what could be causing this unchanging behavior.
> > > Here's my evaluation script:
> > > lm_eval --model vllm
> > > --model_args pretrained=deepseek-ai/DeepSeek-R1-Distill-Llama-8B,
> > > dtype=bfloat16,
> > > tensor_parallel_size=1,
> > > gpu_memory_utilization=0.85,
> > > sae_release=andreuka18/deepseek-r1-distill-llama-8b-lmsys-openthoughts,
> > > sae_id=blocks.19.hook_resid_post,
> > > sae_feature_idx=46379,
> > > sae_strength=2.0,
> > > sae_max_activation=5.840
> > > --tasks aime2024_nofigures
> > > --batch_size auto
> > > --apply_chat_template
> > > --output_path deepseek-llama-8b-aime-sae-46379-test
> > > --log_samples
> > > --gen_kwargs "max_gen_toks=32768,cutoff_token=128014,do_sample=True,temperature=0.6,top_p=0.95"
> > 
> > 
> > Can you clarify what do you mean by "unchanged"? Unchanged compared to what?
> 
> When I run this script multiple times with the same parameters, for each AIME problem, I get exactly identical responses across different runs. Since I've set temperature=0.6 and top_p=0.95 with do_sample=True, I would expect some randomness in the generated responses.

Try to vary the `seed` parameter between different runs. Currently it is fixed for each run, so there indeed should be no randomness between different runs. You can pass it to `--model_args` together with other arguments: `--model_args pretrained=..., dtype=bfloat16,seed=42,...`

The default value is `seed=1234`

**Comment by Peter-Fy** [comment]
> > > > When I run the evaluation, the generated results remain unchanged each time, even though I have explicitly set the temperature and top_p parameters. I would greatly appreciate any insights you might have regarding what could be causing this unchanging behavior.
> > > > Here's my evaluation script:
> > > > lm_eval --model vllm
> > > > --model_args pretrained=deepseek-ai/DeepSeek-R1-Distill-Llama-8B,
> > > > dtype=bfloat16,
> > > > tensor_parallel_size=1,
> > > > gpu_memory_utilization=0.85,
> > > > sae_release=andreuka18/deepseek-r1-distill-llama-8b-lmsys-openthoughts,
> > > > sae_id=blocks.19.hook_resid_post,
> > > > sae_feature_idx=46379,
> > > > sae_strength=2.0,
> > > > sae_max_activation=5.840
> > > > --tasks aime2024_nofigures
> > > > --batch_size auto
> > > > --apply_chat_template
> > > > --output_path deepseek-llama-8b-aime-sae-46379-test
> > > > --log_samples
> > > > --gen_kwargs "max_gen_toks=32768,cutoff_token=128014,do_sample=True,temperature=0.6,top_p=0.95"
> > > 
> > > 
> > > Can you clarify what do you mean by "unchanged"? Unchanged compared to what?
> > 
> > 
> > When I run this script multiple times with the same parameters, for each AIME problem, I get exactly identical responses across different runs. Since I've set temperature=0.6 and top_p=0.95 with do_sample=True, I would expect some randomness in the generated responses.
> 
> Try to vary the `seed` parameter between different runs. Currently it is fixed for each run, so there indeed should be no randomness between different runs. You can pass it to `--model_args` together with other arguments: `--model_args pretrained=..., dtype=bfloat16,seed=42,...`
> 
> The default value is `seed=1234`

Perfect, it's working now. Thanks for the help!
